optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractAccumulator.java

```diff
@@ -85,6 +60,37 @@ public final ResultContainer_ init(Object workingMemoryContext, Object context,
         return containerSupplier.get();
     }
 
+    @Override
+    public final Object accumulate(Object workingMemoryContext, Object context, Tuple leftTuple,
+            InternalFactHandle handle, Declaration[] declarations, Declaration[] innerDeclarations,
+            ReteEvaluator reteEvaluator) {
+        /*
+         * Accumulator instances are created within the KieBase, not within the KieSession.
+         * This means that sessions may be calling this method on a shared accumulator from different threads.
+         * Unfortunately, we need access to innerDeclarations to be able to perform the accumulation quickly,
+         * and it only becomes available when this accumulate(...) method is called.
+         * Therefore, initialization of our accumulator can only happen the first time the method is called,
+         * and therefore the initialization needs to be properly synchronized.
+         *
+         * We do not care if the initialization happens multiple times,
+         * as each of the attempts will fill the variables with the same data.
+         * We just need to ensure that the variables are initialized before the actual accumulation happens.
+         *
+         * Technically running this more than once is wasteful,
+         * but the initialized() check will make sure this only happens during the cold start of the session.
+         */
+        if (!initialized.get()) {
+            initialize(leftTuple, innerDeclarations);
+            initialized.set(true);
+        }
```

Comment:
I fear this might be broken.
Read the entire section "Usage in Java" of https://en.wikipedia.org/wiki/Double-checked_locking

Race condition proof:
- Thread 1 calls accumulate() line 64
- Thread 1 checks if (!initialized.get()) { line 82 => goes to line 83
- Thread 2 calls accumulate() line 64
- Thread 2 checks if (!initialized.get()) { line 82 => goes to line 83
- Thread 1 calls initialize(...) line 83
- Thread 2 calls initialize(...) line 83

Solution: this entire if block should be wrapped by `synchronized(initialized) {`

Reply:
Not that uncontested locks are cheap.

Also it's too back that AtomicBoolean doesn't support something like setIfFalse(Function<Boolean, Boolean>), similar to Maps.putIfAbsent() etc.

Reply:
I am aware of this issue, and I am explaining in the comment why this is not an issue.
Yes, it will cause initialize() to be called multiple times - but we do not care.
Each call to initialize() will set *the same* values to all variables.

The code does what it is supposed to - each thread will see the initialized accumulator.

Reply:
If we want to prevent the multiple initialization, I agree we need to run this in a synchronized block. Is it worth it to force this code to always synchronize, so that we don't have a harmless double call to the method?

Reply:
Non contested synchronization is extremely cheap. It's better to be correct than fast. IIRC Brian Goetz's talk correctly, we shouldn't try to be a smartass and just synchronize it.

But even better would be if we can get Drools to allow us to just use the init() method for this work.